README.md

```diff
@@ -1,16 +1,72 @@
+
+For example:
+
+- if engineering decides to migrate to GraphQL instead of a REST API, only the data layer will be impacted
+- if product wants to change a criteria for the business, only the domain layer will be impacted
+- if design wants to change the application theme, only the presentation layer will be impacted
+
+### Layers
+
+![bloc_architecture_diagram](https://user-images.githubusercontent.com/29372054/178124498-6675ede0-d0ea-40eb-98dc-136991af6b28.png)
+
+#### Data Layer
+
+This layer is the lowest layer and is responsible for retrieving raw data from external sources(database, REST API, GraphQL backend) or device APIs. Usually packages in this layer will expose `clients` (e.g.: `storage_client`, `location_client`, `auth0_api_client`, etc...).
+
+- Packages in the data layer should not import any Flutter dependencies.
+- Clients can be reused and even published on pub.dev as standalone packages.
+- Packages in the data layer should not be dependent on other clients.
+- This layer should not interact with any other layers.
+- A single client should be created per data source.
+
+**Note**: This layer can be considered the "engineering" layer because it focuses on how to process/transform data in a performant way.
+
+#### Domain Layer
+
+This layer is a compositional layer meaning that it composes one or more data clients and applies "business rules" to the data. We call each component in this layer a `repository` (e.g.: `user_repository`, `weather_repository`, `payments_repository`, etc...).
+
+- Repositories should not import any Flutter dependencies.
+- Packages in the repository layer should not be dependent on other repositories.
+- This layer should only interact with the data layer.
+- One repository should be created per domain model.
+
+**Note**: This layer can be considered the "product" layer. The business/product owner will determine the rules/acceptance criteria for how to combine data from one or more data providers into a unit that brings value to the customer.
+
+#### Business Logic Layer
+
+This layer composes one or more repositories and contains logic for how to surface the business rules via a specific feature or use-case. This layer uses the [bloc package](https://pub.dev/packages/bloc) to manage the logic associated with each feature. We call each component in this layer a `bloc` (e.g.: `login_bloc`, `weather_forecast_bloc`, `settings_bloc`, etc...).
+
+- The business logic layer should have no dependency on the Flutter SDK
+- The business logic layer should not have direct dependencies on other business logic components.
+- This layer should only interact with the domain layer.
+- One bloc/cubit should be created per feature(see also [cubit vs bloc](https://bloclibrary.dev/#/coreconcepts?id=cubit-vs-bloc)).
```

Comment:
How will it be decided if a cubit or bloc is appropriate? I see that cubits are more simple implementations... Will we say that we should defer to using cubits unless a bloc is absolutely needed?

Reply:
Indeed cubits are simpler to use and are great when you don't care about tracing events or manipulating app events.
If that's something you care about then bloc is the way to go.

Say you have an authentication feature, you might wanna go with a bloc to trace exactly what triggers authenticated & unauthenticated states.
Or say you have a feature which contains a search or typeahead, then you might also go with a bloc because you want to be able to easily debounce the incoming search phrase so that you avoid unnecessary backend requests.

If you're unsure what to do it's simpler to start with a cubit then migrate to a bloc when such scenarios arise.

Reply:
I think that's fine with me.